### PR TITLE
Add CMSSW exception message to condor chirp, when present

### LINKED ISSUE #6939: Chirp exception message
We've had sites who want to explore the ClassAd archive based on a full text search of exception messages.

As part of the WMCore wrapper that currently parses & chirps back the FJR, we should also include the exception message for jobs that fail.

However, with the exception messages, we are likely to hit the 1024 character limit on values passed via Chirp.  Hence, we should extract the exception message, compress it and base64-encode it.  If the encoded version is still greater than 1024 character, then we should recursively repeat the procedure until the encoded text is below the 1024 character threshold.  ElasticSearch can decode things appropriately on the other side.

### PR DESCRIPTION
**Pull request by khurtado**
Fixes #6939

### CODE REVIEW

**Comment by khurtado** [comment]
@amaltaro This seems to be working (tested on vocms0192) [1]
Message is encoded and compressed.
For reference, the decoding/decompression need to be done at the ES side, here is the PR for that:
https://github.com/dmwm/cms-htcondor-es/pull/100

[1] If there is an error message, it will be passed via chirp as `Chirp_WMCore_<>_Exception_Message`
```
[khurtado@vocms0192 ~]$ condor_history -limit 100 -constraint '!isUndefined(Chirp_WMCore_cmsRun1_ExitCode)' -af:h Chirp_WMCore_cmsRun1_ExitCode Chirp_WMCore_cmsRun1_Exception_Message
Chirp_WMCore_cmsRun1_ExitCode Chirp_WMCore_cmsRun1_Exception_Message
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
0                             undefined                             
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
8001                          eNpdj7tqw1AQRHt9xeBGXZAgTQQukli4CpjYRSAEsVqtpAvXu+E+HPvvoziPIu3MYThzr5Azy3typrARTEkmCxeUj6aD+4o3lKinKCWMOYcgAz5m56UA8Fq9YQFjCpmT0wlpFrQn0bQLxhKjhStW/8Pa/d5yYGnAnmJc78z85uE3hade/LrceuvJH2gqi/ZP8mlZpUma4rvFUmN1V710R37OWndByLuYHHen27paYaYItYReRDFa1gFOr57Dz7EbjMGO2B4W5/OlabzR8AlOdmJG
```

File: src/python/WMCore/WMSpec/Steps/Executor.py [hunk 1]
```diff
@@ -161,14 +169,16 @@ def post(self, emulator = None):
...
-
-    def setCondorChirpAttrDelayed(self, key, value):
+    def setCondorChirpAttrDelayed(self, key, value, compress=False, maxLen=5120):
         """
         _setCondorChirpAttrDelayed_
 
         Util to call condor_chirp and publish the key/value pair
 
         """
+        if compress == True:
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Executor.py]: `if compress is True:` or just `if compress:`

File: src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py [hunk 2]
```diff
@@ -42,13 +42,16 @@ class CMSSW(Executor):
...
+    def _setStatus(self, returnCode, returnMessage):
         """
         Set return code.
         """
-        self.setCondorChirpAttrDelayed('Chirp_WMCore_cmsRun_ExitCode', returncode)
-        self.setCondorChirpAttrDelayed('Chirp_WMCore_%s_ExitCode' % self.stepName, returncode)
-        self.step.execution.exitStatus = returncode
+        self.setCondorChirpAttrDelayed('Chirp_WMCore_cmsRun_ExitCode', returnCode)
+        self.setCondorChirpAttrDelayed('Chirp_WMCore_%s_ExitCode' % self.stepName, returnCode)
+        if returnMessage:
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py]: should we also check whether `returnCode != 0`? Such that we don't even chirp that attribute if the job didn't fail?

File: src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py [hunk 3]
```diff
@@ -241,24 +244,26 @@ def execute(self, emulator=None):
...
 
-        if returncode != 0:
+        if returnCode != 0:
             argsDump = {'arguments': args}
             msg = "Error running cmsRun\n%s\n" % argsDump
             try:
                 self.report.parse(jobReportXML, stepName=self.stepName)
-                returncode = self.report.getStepExitCode(stepName=self.stepName)
-                msg += "CMSSW Return code: %s\n" % returncode
+                (returnCode, returnMessage) = \
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py]: Sometimes I prefer not to respect the 80 chars line length, just so the code is easier to read. This seems to be a good one for that :-)

**Review by amaltaro — CHANGES_REQUESTED** [review]
Patch looks good in general, Kenyi.
I made a few comments along the code, mostly minor suggestions though.

I was going to suggest you to add another function for decode + decompressing it, but if we won't use it inside WMCore, then I think we can just forget it.

About code reformatting and pylint changes (+ unit tests), can you please keep them in a separate commit for future PRs? That way we can skip those changes and look only at the 1st commit (the one with real changes).

**Comment by amaltaro** [comment]
test this please

File: src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py (hunk 2 above)
Comment (khurtado) [reply src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py]: I think the Report will always return a `None` return message if exit code is 0, but an extra safety check doesn't hurt.

File: src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py [hunk 4]
```diff
@@ -241,24 +244,25 @@ def execute(self, emulator=None):
...
-                returncode = self.report.getStepExitCode(stepName=self.stepName)
-                msg += "CMSSW Return code: %s\n" % returncode
+                (returnCode, returnMessage) = self.report.getStepExitCodeAndMessage(stepName=self.stepName)
+                msg += "CMSSW Return code: %s\n" % returnCode
             except Exception as ex:
                 # If report parsing fails, report linux exit code
-                msg += "Linux Return code: %s\n" % returncode
+                msg += "Linux Return code: %s\n" % returnCode
             finally:
                 logging.critical(msg)
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py]: How about adding a `logging.critical("Error message: %s", returnMessage)` in the next line?
Sorry for not suggesting it in the previous review.

**Review by amaltaro — COMMENTED** [review]
Patch looks good. Can you just add that extra line I commented in the code? Have you tested this again after my previous review?

File: src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py (hunk 4 above)
Comment (khurtado) [reply src/python/WMCore/WMSpec/Steps/Executors/CMSSW.py]: Done

**Comment by khurtado** [comment]
> Patch looks good. Can you just add that extra line I commented in the code? Have you tested this again after my previous review?

@amaltaro Extra line added. I just patched the Agent in `vocms0192` and am running a new test to double check.

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
Let's forget that unit test failure. Issue created to get it fixed eventually:
https://github.com/dmwm/WMCore/issues/9071

**Review by amaltaro — APPROVED** [review]
Patch looks good, Kenyi. Just let me know the outcome of your current tests and we can merge it.

**Comment by khurtado** [comment]
@amaltaro : Performance tests gave reasonable results and changes have been merged in the CMS/ES  [https://github.com/dmwm/cms-htcondor-es/pull/100](repository).

**Comment by amaltaro** [comment]
Awesome, thanks!

**Pull request merged**